### ISSUE TITLE
SharedArrayBuffer content drive not working

### ISSUE THREAD

**Issue by juntyr** [jupyterlite/jupyterlite#1429]
I wanted to try out the new SharedArrayBuffer-based content sharing for my deploy of JupyterLite. In my requirements.txt, I set

```
jupyterlite-core==0.4.0b1
jupyterlab~=4.2.3
jupyterlite-pyodide-kernel==0.4.0b0
```

My previous jupyter-lite.json config disabled the service worker and only used in-memory storage

```json
{
    "jupyter-lite-schema-version": 0,
    "jupyter-config-data": {
        "disabledExtensions": ["@jupyterlite/server-extension:service-worker"],
        "enableMemoryStorage": true,
        "contentsStorageDrivers": ["memoryStorageDriver"]
    }
}
```

Since I'm also using the SharedArrayBuffer-based trick for synchronous communication, my site is already deployed with the required HTTP CORS headers.

Still, when using the site, I get the following warnings:

```
JupyterLite ServiceWorker not available [index.js:185:24]
Kernel filesystem and JupyterLite contents will NOT be synced [index.js:188:24]
```

and the `/drive` mount is missing:

```python
from pathlib import Path
list(Path("/").iterdir())

# [PosixPath('/tmp'),
#  PosixPath('/home'),
#  PosixPath('/dev'),
#  PosixPath('/proc'),
#  PosixPath('/lib'),
#  PosixPath('/usr')]
```

Is there anything else I need to change in my settings to enable the new SharedArrayBuffer-based content sharing while also continuing to not enable the service worker?

Thanks for your help!

_Originally posted by @juntyr in https://github.com/jupyterlite/jupyterlite/issues/1405#issuecomment-2205851769_

**Comment by gautamsarawagi** [comment]
Hey @juntyr getting this same issue. As soon as I allow the third-party-cookies it works just as fine. But without it gives the same error. I also doubt this to be the same issue. 

![image](https://github.com/user-attachments/assets/e39326b9-5fa5-4ef7-b6ae-c971730bb6a2)


Were you able to solve it?

**Comment by juntyr** [comment]
@jtpio It seems the Pyodide kernel is at fault - in https://github.com/jupyterlite/pyodide-kernel/blob/main/packages/pyodide-kernel-extension/src/index.ts in the create function, the drive mount is only enabled if the service worker is enabled, even though this is now not necessary. Perhaps this check could be adapted to alternatively allow the drive to be mounted if the shared buffer implementation will be used?

**Comment by gautamsarawagi** [comment]
@juntyr @jtpio 

Also got to know about this from the doc: https://jupyterlite.readthedocs.io/en/stable/howto/content/python.html

but diesn;t get executed as the cookies are blocked: 

![image](https://github.com/user-attachments/assets/84f265f0-d216-4290-9f8d-90cfa2793700)


otherwise in the browser where the third party cokkies are enabled. It works just as fine: 

![image](https://github.com/user-attachments/assets/43b0073c-b32f-487b-85a0-39cd3dd8fe3c)

**Comment by juntyr** [comment]
@gautamsarawagi Thank you for sharing your problem, though I think you're looking at a different issue. In my case, the service worker is always disabled but SharedArrayBuffers are guaranteed to work, so the drive sync should work but currently doesn't because of the incorrect check in https://github.com/jupyterlite/pyodide-kernel/blob/7e9351258ef41a9f2dfdd96753ea2ee1e719d12b/packages/pyodide-kernel-extension/src/index.ts#L82.

**Comment by gautamsarawagi** [comment]
Thanks @juntyr I totally got the point and yes the problem which I am facing is related to the service-worker that might gets resolved with the use of `SharedArrayBuffers`.

**Comment by jtpio** [comment]
> @jtpio It seems the Pyodide kernel is at fault - in https://github.com/jupyterlite/pyodide-kernel/blob/main/packages/pyodide-kernel-extension/src/index.ts in the create function, the drive mount is only enabled if the service worker is enabled, even though this is now not necessary. Perhaps this check could be adapted to alternatively allow the drive to be mounted if the shared buffer implementation will be used?

Oh, it looks like this should indeed not be the case. Maybe we should check for `crossOriginIsolated` there too.

**Comment by jtpio** [comment]
@juntyr would you be able to try the built version of `jupyterlite-pyodide-kernel` from this PR? https://github.com/jupyterlite/pyodide-kernel/pull/129

Available in the GitHub artifact: https://github.com/jupyterlite/pyodide-kernel/actions/runs/10036926748?pr=129

**Comment by juntyr** [comment]
> @juntyr would you be able to try the built version of `jupyterlite-pyodide-kernel` from this PR? [jupyterlite/pyodide-kernel#129](https://github.com/jupyterlite/pyodide-kernel/pull/129)
> 
> Available in the GitHub artifact: https://github.com/jupyterlite/pyodide-kernel/actions/runs/10036926748?pr=129

Yes it works!

**Comment by jtpio** [comment]
Great, thanks for testing!

**Comment by gautamsarawagi** [comment]
@jtpio The above mounting issue really solves the issue of running the kernel in different browsers and the contetns also geting synced. 

But the only problem that I am experincing currently is that it doesn;t work well with the **iframes**.

The iframe I am using: 

 ```
<iframe
                            id="jupyterIframe"
                            src={'https://funny-druid-02ae7d.netlify.app/lite/lab/index.html'}
                            width="100%"
                            height="100%"
                            style={{
                                border: 'none',
                                boxShadow: '0px 0px 10px rgba(0, 0, 0, 0.1)',
                                borderRadius: '8px',
                            }}        
                            allow="cross-origin-isolated"
                             ></iframe>
```

and the jupyter-lite.json:

```
"appName": "JupyterLite",
    "appUrl": "./lab",
    "appVersion": "0.4.0-rc.0",
    "baseUrl": "./",
    "contentsStorageDrivers": [
      "memoryStorageDriver",
      "indexedDB"
    ],
    "defaultKernelName": "python",
    "disabledExtensions": [
      "@jupyterlite/server-extension:service-worker"
    ],
    "enableMemoryStorage": true,
    "exposeAppInBrowser": true,
    "faviconUrl": "./lab/favicon.ico",
    "federated_extensions": [
      {
        "extension": "./extension",
        "liteExtension": true,
        "load": "static/remoteEntry.cd2230d4fe3113507a18.js",
        "name": "@jupyterlite/pyodide-kernel-extension"
      },
      {
        "extension": "./extension",
        "liteExtension": false,
        "load": "static/remoteEntry.56c576fbaee147c549ef.js",
        "name": "cbioportal-jupyterlite",
        "style": "./style"
      },
      {
        "extension": "./extension",
        "liteExtension": false,
        "load": "static/remoteEntry.5cbb9d2323598fbda535.js",
        "name": "jupyterlab_pygments",
        "style": "./style"
      }
    ],
```

**Comment by jtpio** [comment]
> src={'https://funny-druid-02ae7d.netlify.app/lite/lab/index.html'}

@gautamsarawagi do you have another example URL? This one seems to be giving a 404:

![image](https://github.com/user-attachments/assets/1ad2f182-9944-4b2d-b80a-fae59d4340c9)

**Comment by jtpio** [comment]
> But the only problem that I am experincing currently is that it doesn;t work well with the **iframes**.

The `allow="cross-origin-isolated"` looks correct. Could it be because the host page should also be served with the COOP / COEP headers?

Testing locally with the following directory structure:

```
index.html
jupyter-lite.json
_output
```

With the following `index.html`:

```html
<iframe
  id="jupyterIframe"
  src="http://localhost:8080"
  width="100%"
  height="100%"
  style={{
  border: 'none',
  boxShadow: '0px 0px 10px rgba(0, 0, 0, 0.1)',
  borderRadius: '8px',
  allow="cross-origin-isolated"
  }}        
></iframe>
```

And serving:

- the main page with `npx static-handler --cors --coop --coep --corp ./`
- the JupyterLite site with `npx static-handler --port 8081 --cors --coop --coep --corp ./_output`

It looks like the IFrame is `crossOriginIsolated`:

![image](https://github.com/user-attachments/assets/788e4546-e222-45d2-820c-eb870950e8a8)

**Comment by gautamsarawagi** [comment]
Yes actually for using the `SharedArrayBuffer` I suppose the host page should also be served with the `COOP / COEP `headers.

**Comment by gautamsarawagi** [comment]
> * the JupyterLite site with `npx static-handler --port 8081 --cors --coop --coep --corp ./_output`

@jtpio I have actually served the jupyterlite deployed url with the cors headers required. 

netllify.toml file:

```
[[headers]]
for = "/*"
  [headers.values]
    Cross-Origin-Opener-Policy = "same-origin"
    Cross-Origin-Embedder-Policy = "require-corp"
```

One more thing I noticed that if you ar**not disabling the `@jupyterlite/server-extension:service-worker` **the contents gets mounted. But to run the kernel effectively we actually need to disable that one.

which could be beacuse of this: 

![image](https://github.com/user-attachments/assets/529fdf43-8120-4e49-b6a5-83a322722c38)

**Comment by jtpio** [comment]
> I have actually served the jupyterlite deployed url with the cors headers required.

Yes, but is the host page also served with the right headers?

> One more thing I noticed that if you ar**not disabling the @jupyterlite/server-extension:service-worker **the contents gets mounted. But to run the kernel effectively we actually need to disable that one.

Are you using the latest pre-release of `jupyterlite-pyodide-kernel`? This should have been fixed by https://github.com/jupyterlite/pyodide-kernel/pull/129.

**Comment by gautamsarawagi** [comment]
> Are you using the latest pre-release of `jupyterlite-pyodide-kernel`? This should have been fixed by [jupyterlite/pyodide-kernel#129](https://github.com/jupyterlite/pyodide-kernel/pull/129).

Yes I am using the pre-released version: 

![image](https://github.com/user-attachments/assets/b4f0e03d-4aaa-4ab4-8706-68ac0259f1eb)

**Comment by jtpio** [comment]
The fix is available in `0.4.0rc0` though, so you may need to update it: https://github.com/jupyterlite/pyodide-kernel/releases/tag/v0.4.0rc0
